File: src/m/classes/qmu/nonlinear_equality_constraint.m
%
%  constructor for the nonlinear_equality_constraint class.
%
%  [nec]=nonlinear_equality_constraint(varargin)
%
%  where the required varargin are:
%    descriptor    (char, description, '')
%    target        (double, target value, 0.)
%  and the optional varargin and defaults are:
%    scale_type    (char, scaling type, 'none')
%    scale         (double, scaling factor, 1.)
%
%  note that zero arguments constructs a default instance; one
%  argument of the class copies the instance; and two or more
%  arguments constructs a new instance from the arguments.
%
%  "Copyright 2009, by the California Institute of Technology.
%  ALL RIGHTS RESERVED. United States Government Sponsorship
%  acknowledged. Any commercial use must be negotiated with
%  the Office of Technology Transfer at the California Institute
%  of Technology.  (J. Schiermeier, NTR 47078)
%
%  This software may be subject to U.S. export control laws.
%  By accepting this  software, the user agrees to comply with
%  all applicable U.S. export laws and regulations. User has the
%  responsibility to obtain export licenses, or other export
%  authority as may be required before exporting such information
%  to foreign countries or providing access to foreign persons."
%
classdef nonlinear_equality_constraint
    properties
        descriptor='';
        target    = 0.;
        scale_type='none';
        scale     = 1.;
    end

    methods
        function [nec]=nonlinear_equality_constraint(varargin)

            switch nargin

%  create a default object

                case 0

%  copy the object

                case 1
                    if isa(varargin{1},'nonlinear_equality_constraint')
                        nec=varargin{1};
                    else
                        error('Object ''%s'' is a ''%s'' class object, not ''%s''.',...
                            inputname(1),class(varargin{1}),'nonlinear_equality_constraint');
                    end

%  create the object from the input

                otherwise
                    asizec=num2cell(array_size(varargin{1:min(nargin,4)}));
                    nec(asizec{:})=nonlinear_equality_constraint;
                    clear asizec

                    if ischar(varargin{1})
                        varargin{1}=cellstr(varargin{1});
                    end
                    for i=1:numel(nec)
                        if (numel(varargin{1}) > 1)
                            nec(i).descriptor=varargin{1}{i};
                        else
                            nec(i).descriptor=[char(varargin{1}) string_dim(nec,i,'vector')];
                        end
                        if (numel(varargin{2}) > 1)
                            nec(i).target    =varargin{2}(i);
                        else
                            nec(i).target    =varargin{2};
                        end
                    end

                    if (nargin >= 3)
                        if ischar(varargin{3})
                            varargin{3}=cellstr(varargin{3});
                        end
                        for i=1:numel(nec)
                            if (numel(varargin{3}) > 1)
                                nec(i).scale_type=varargin{3}{i};
                            else
                                nec(i).scale_type=char(varargin{3});
                            end
                        end
                        if (nargin >= 4)
                            for i=1:numel(nec)
                                if (numel(varargin{4}) > 1)
                                    nec(i).scale     =varargin{4}(i);
                                else
                                    nec(i).scale     =varargin{4};
                                end
                            end

                            if (nargin > 4)
                                warning('objective_function:extra_arg',...
                                    'Extra arguments for object of class ''%s''.',...
                                    class(nec));
                            end
                        end
                    end
            end

        end

        function []=disp(nec)

%  display the object

            disp(sprintf('\n'));
            for i=1:numel(nec)
                disp(sprintf('class ''%s'' object ''%s%s'' = \n',...
                    class(nec),inputname(1),string_dim(nec,i)));
                disp(sprintf('    descriptor: ''%s'''  ,nec(i).descriptor));
                disp(sprintf('        target: %g'      ,nec(i).target));
                disp(sprintf('    scale_type: ''%s'''  ,nec(i).scale_type));
                disp(sprintf('         scale: %g\n'    ,nec(i).scale));
            end

        end

        function [desc]  =prop_desc(nec,dstr)
            desc=cell(1,numel(nec));
            for i=1:numel(nec)
                if ~isempty(nec(i).descriptor)
                    desc(i)=cellstr(nec(i).descriptor);
                elseif ~isempty(inputname(1))
                    desc(i)=cellstr([inputname(1) string_dim(nec,i,'vector')]);
                elseif exist('dstr','var')
                    desc(i)=cellstr([dstr         string_dim(nec,i,'vector')]);
                else
                    desc(i)=cellstr(['nec'        string_dim(nec,i,'vector')]);
                end
            end
            desc=allempty(desc);
        end
        function [stype] =prop_stype(nec)
            stype=cell(size(nec));
            for i=1:numel(nec)
                stype(i)=cellstr(nec(i).scale_type);
            end
            stype=allequal(stype,'none');
        end
        function [scale] =prop_scale(nec)
            scale=zeros(size(nec));
            for i=1:numel(nec)
                scale(i)=nec(i).scale;
            end
            scale=allequal(scale,1.);
        end
        function [weight]=prop_weight(nec)
            weight=[];
        end
        function [lower] =prop_lower(nec)
            lower=[];
        end
        function [upper] =prop_upper(nec)
            upper=[];
        end
        function [target]=prop_target(nec)
            target=zeros(size(nec));
            for i=1:numel(nec)
                target(i)=nec(i).target;
            end
            target=allequal(target,0.);
        end
    end

    methods (Static)
        function [rdesc]=dakota_write(fidi,dresp,rdesc)

%  collect only the responses of the appropriate class

            nec=struc_class(dresp,'nonlinear_equality_constraint');

%  write responses

            [rdesc]=rlist_write(fidi,'nonlinear_equality_constraints','nonlinear_equality',nec,rdesc);
        end

        function []=dakota_rlev_write(fidi,dresp,params)
        end
    end
end

File: src/m/classes/qmu/nonlinear_equality_constraint.py
import numpy as np
from rlist_write import *
from MatlabArray import *


class nonlinear_equality_constraint:
    '''
  constructor for the nonlinear_equality_constraint class.

  [nec] = nonlinear_equality_constraint.nonlinear_equality_constraint(args)
   nec = nonlinear_equality_constraint()

  where the required args are:
    descriptor    (char, description, '')
    target        (double vector, target values, 0.)
  and the optional args and defaults are:
    scale_type    (char, scaling type, 'none')
    scale         (double, scaling factor, 1.)

  note that zero arguments constructs a default instance, one
  argument of the class copies the instance, and two or more
  arguments constructs a new instance from the arguments.
'''
    def __init__(self):
        self.descriptor = ''
        self.target = 0.
        self.scale_type = 'none'
        self.scale = 1.

    @staticmethod
    def nonlinear_equality_constraint(*args):
        nargin = len(args)

    # create a default object
        if nargin == 0:
            return nonlinear_equality_constraint()

    # copy the object
        elif nargin == 1:
            if isinstance(args[0], nonlinear_equality_constraint):
                nec = args[0]
            else:
                raise RuntimeError('Object is a ' + str(type(args[0])) + ' class object, not "nonlinear_equality_constraint"')

    # create the object from the input
        else:
            asizec = array_size(*args[0:min(nargin, 4)])
            nec = [nonlinear_equality_constraint() for i in range(asizec[0]) for j in range(asizec[1])]

            for i in range(np.size(nec)):
                if (np.shape(args[0])[0] > 1):
                    nec[i].descriptor = args[0][i]
                else:
                    nec[i].descriptor = str(args[0])
                if (np.size(args[1]) > 1):
                    nec[i].target = args[1][i]
                else:
                    nec[i].target = args[1]

            if (nargin >= 3):
                for i in range(np.size(nec)):
                    if (np.size(args[2]) > 1):
                        nec[i].scale_type = args[2][i]
                    else:
                        nec[i].scale_type = str(args[2])

            if (nargin >= 4):
                for i in range(np.size(nec)):
                    if (np.size(args[3]) > 1):
                        nec[i].scale = args[3][i]
                    else:
                        nec[i].scale = args[3]

            if (nargin > 4):
                print('WARNING: nonlinear_equality_constraint:extra_arg: Extra arguments for object of class ' + str(type(nec)) + '.')

        return nec

    def __repr__(self):
        # display the object
        string = '\n'
        string += 'class "nonlinear_equality_constraint" object = \n'
        string += '    descriptor: ' + str(self.descriptor) + '\n'
        string += '        target: ' + str(self.target) + '\n'
        string += '    scale_type: ' + str(self.scale_type) + '\n'
        string += '         scale: ' + str(self.scale) + '\n'

        return string

    @staticmethod
    def prop_desc(nec, dstr):
        if type(nec) not in [list, np.ndarray]:
            if nec.descriptor != '' or type(nec.descriptor) != str:
                desc = str(nec.descriptor)
            elif dstr != '':
                desc = str(dstr)
            else:
                desc = 'nec'
            return desc

        desc = ['' for i in range(np.size(nec))]
        for i in range(np.size(nec)):
            if nec[i].descriptor != '' or type(nec[i].descriptor) != str:
                desc[i] = str(nec[i].descriptor)
            elif dstr != '':
                desc[i] = str(dstr) + str(string_dim(nec, i, 'vector'))
            else:
                desc[i] = 'nec' + str(string_dim(nec, i, 'vector'))

        desc = allempty(desc)

        return desc

    @staticmethod
    def prop_lower(nec):
        lower = []
        return lower

    @staticmethod
    def prop_upper(nec):
        upper = []
        return upper

    @staticmethod
    def prop_weight(nec):
        weight = []
        return weight

    @staticmethod
    def prop_target(nec):
        if type(nec) not in [list, np.ndarray]:
            return nec.target

        target = np.zeros(np.shape(nec))
        for i in range(np.size(nec)):
            target[i] = nec[i].target

        target = allequal(target, 0.)

        return target

    @staticmethod
    def prop_stype(nec):
        if type(nec) not in [list, np.ndarray]:
            return nec.scale_type

        stype = ['' for i in range(np.size(nec))]
        for i in range(np.size(nec)):
            stype[i] = str(nec[i].scale_type)

        stype = allequal(stype, 'none')

        return stype

    @staticmethod
    def prop_scale(nec):
        if type(nec) not in [list, np.ndarray]:
            return nec.scale

        scale = np.zeros(np.shape(nec))
        for i in range(np.size(nec)):
            scale[i] = nec[i].scale

        scale = allequal(scale, 1.)

        return scale

    @staticmethod
    def dakota_write(fidi, dresp, rdesc):
        #  colnect only the variables of the appropriate class
        nec = [struc_type(i, 'nonlinear_equality_constraint', 'nec') for i in dresp]

        #  write constraints
        rdesc = rlist_write(fidi, 'nonlinear_equality_constraints', 'nonlinear_equality', nec, rdesc)
        return rdesc

    @staticmethod
    def dakota_rlev_write(fidi, dresp, params):
        return
